File: src/connect4_zero/search/tree.py
"""Explicit tree objects for batched deep MCTS."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Tuple

import torch

from connect4_zero.game import Connect4x4x4Batch, StepResult
from connect4_zero.game.constants import ACTION_SIZE


@dataclass
class TreeNode:
    """One MCTS node, valued from this node's player-to-move perspective."""

    state: Connect4x4x4Batch
    parent: Optional["TreeNode"]
    parent_action: Optional[int]
    legal_actions: Tuple[int, ...]
    terminal_value: Optional[float] = None
    visits: int = 0
    value_sum: float = 0.0
    virtual_visits: int = 0
    virtual_value_sum: float = 0.0
    depth: int = 0
    children: List[Optional["TreeNode"]] = field(
        default_factory=lambda: [None for _ in range(ACTION_SIZE)]
    )

    @property
    def is_terminal(self) -> bool:
        return self.terminal_value is not None

    @property
    def mean_value(self) -> float:
        if self.visits == 0:
            return 0.0
        return self.value_sum / self.visits

    @property
    def effective_visits(self) -> int:
        return self.visits + self.virtual_visits

    @property
    def effective_mean_value(self) -> float:
        visits = self.effective_visits
        if visits == 0:
            return 0.0
        return (self.value_sum + self.virtual_value_sum) / visits

    @property
    def is_fully_expanded(self) -> bool:
        return all(self.children[action] is not None for action in self.legal_actions)

    def unexpanded_actions(self) -> Tuple[int, ...]:
        return tuple(action for action in self.legal_actions if self.children[action] is None)


@dataclass
class SearchTree:
    """A single-root game tree for one current self-play position."""

    root: TreeNode
    num_nodes: int = 1
    max_depth: int = 0

    @classmethod
    def from_state(cls, state: Connect4x4x4Batch) -> "SearchTree":
        """Create a tree whose root is a cloned single-state batch."""
        _validate_single_state(state)
        root_state = state.clone()
        terminal_value = terminal_value_for_state(root_state)
        root = TreeNode(
            state=root_state,
            parent=None,
            parent_action=None,
            legal_actions=legal_actions_for_state(root_state),
            terminal_value=terminal_value,
        )
        return cls(root=root)

    def expand_child(self, parent: TreeNode, action: int) -> TreeNode:
        """Expand ``parent/action`` and return the resulting child node."""
        if parent.is_terminal:
            raise RuntimeError("cannot expand a terminal node")
        if action not in parent.legal_actions:
            raise RuntimeError(f"cannot expand illegal action {action}")

        existing = parent.children[action]
        if existing is not None:
            return existing

        child_state = parent.state.clone()
        result = child_state.step(torch.tensor([action], dtype=torch.long, device=child_state.device))
        if not bool(result.legal[0].item()):
            raise RuntimeError(f"engine rejected legal action {action}")

        child = TreeNode(
            state=child_state,
            parent=parent,
            parent_action=action,
            legal_actions=legal_actions_for_state(child_state),
            terminal_value=terminal_value_for_child(result),
            depth=parent.depth + 1,
        )
        parent.children[action] = child
        self.num_nodes += 1
        self.max_depth = max(self.max_depth, child.depth)
        return child

    def reuse_child(self, action: int) -> Optional["SearchTree"]:
        """Return a new tree rooted at the existing child for ``action``."""
        if action < 0 or action >= ACTION_SIZE:
            return None

        child = self.root.children[action]
        if child is None:
            return None
        return self.from_existing_root(child)

    @classmethod
    def from_existing_root(cls, root: TreeNode) -> "SearchTree":
        """Detach ``root`` from its parent and rebase subtree depths."""
        root.parent = None
        root.parent_action = None
        num_nodes, max_depth = _rebase_and_count(root, depth=0)
        return cls(root=root, num_nodes=num_nodes, max_depth=max_depth)


def legal_actions_for_state(state: Connect4x4x4Batch) -> Tuple[int, ...]:
    """Return legal action indices for a single-state batch."""
    _validate_single_state(state)
    return tuple(
        int(action)
        for action in state.legal_mask()[0].nonzero(as_tuple=False).flatten().detach().cpu().tolist()
    )


def terminal_value_for_state(state: Connect4x4x4Batch) -> Optional[float]:
    """Return exact value for a terminal root, from player-to-move perspective."""
    _validate_single_state(state)
    if bool(state.done[0].item()):
        return -1.0 if int(state.outcome[0].item()) == 1 else 0.0
    if not bool(state.legal_mask()[0].any().item()):
        return 0.0
    return None


def terminal_value_for_child(result: StepResult) -> Optional[float]:
    """Return exact child value after a one-state step, if terminal."""
    if bool(result.won[0].item()):
        return -1.0
    if bool(result.draw[0].item()):
        return 0.0
    return None


def _validate_single_state(state: Connect4x4x4Batch) -> None:
    if state.batch_size != 1:
        raise ValueError(f"tree nodes require batch_size=1, got {state.batch_size}")


def _rebase_and_count(node: TreeNode, depth: int) -> Tuple[int, int]:
    node.depth = depth
    num_nodes = 1
    max_depth = depth
    for child in node.children:
        if child is None:
            continue
        child.parent = node
        child_count, child_max_depth = _rebase_and_count(child, depth=depth + 1)
        num_nodes += child_count
        max_depth = max(max_depth, child_max_depth)
    return num_nodes, max_depth

File: archive/rust_prototype_2026_05_14/connect4_zero_rust/crates/c4-search/src/tree.rs
use c4_core::{Action, Position, constants::ACTION_COUNT};

pub type NodeId = usize;

#[derive(Clone, Debug)]
pub struct Node {
    pub position: Position,
    pub parent: Option<NodeId>,
    pub parent_action: Option<Action>,
    pub children: [Option<NodeId>; ACTION_COUNT],
    pub legal_mask: u16,
    pub visits: u32,
    pub value_sum: f32,
    pub prior: f32,
    pub terminal_value: Option<f32>,
    pub priors: [f32; ACTION_COUNT],
    pub expanded: bool,
    pub depth: usize,
}

impl Node {
    pub fn root(position: Position) -> Self {
        Self {
            position,
            parent: None,
            parent_action: None,
            children: [None; ACTION_COUNT],
            legal_mask: position.legal_mask(),
            visits: 0,
            value_sum: 0.0,
            prior: 1.0,
            terminal_value: position.terminal_value(),
            priors: [0.0; ACTION_COUNT],
            expanded: false,
            depth: 0,
        }
    }

    pub fn mean_value(&self) -> f32 {
        if self.visits == 0 {
            0.0
        } else {
            self.value_sum / self.visits as f32
        }
    }
}

#[derive(Clone, Debug)]
pub struct SearchTree {
    pub nodes: Vec<Node>,
    pub root: NodeId,
    pub reused: bool,
}

impl SearchTree {
    pub fn new(position: Position) -> Self {
        Self {
            nodes: vec![Node::root(position)],
            root: 0,
            reused: false,
        }
    }

    pub fn root(&self) -> &Node {
        &self.nodes[self.root]
    }

    pub fn root_mut(&mut self) -> &mut Node {
        &mut self.nodes[self.root]
    }

    pub fn add_child(&mut self, parent: NodeId, action: Action, position: Position) -> NodeId {
        if let Some(child) = self.nodes[parent].children[action.index()] {
            return child;
        }
        let depth = self.nodes[parent].depth + 1;
        let child = Node {
            position,
            parent: Some(parent),
            parent_action: Some(action),
            children: [None; ACTION_COUNT],
            legal_mask: position.legal_mask(),
            visits: 0,
            value_sum: 0.0,
            prior: self.nodes[parent].priors[action.index()],
            terminal_value: position.terminal_value(),
            priors: [0.0; ACTION_COUNT],
            expanded: false,
            depth,
        };
        let child_id = self.nodes.len();
        self.nodes.push(child);
        self.nodes[parent].children[action.index()] = Some(child_id);
        child_id
    }

    pub fn advance_to_child(&mut self, action: Action) -> bool {
        match self.nodes[self.root].children[action.index()] {
            Some(child) => {
                self.root = child;
                self.reused = true;
                true
            }
            None => false,
        }
    }

    pub fn max_depth_from_root(&self) -> usize {
        self.nodes
            .iter()
            .filter(|node| self.is_descendant_of_root(node))
            .map(|node| node.depth.saturating_sub(self.nodes[self.root].depth))
            .max()
            .unwrap_or(0)
    }

    fn is_descendant_of_root(&self, node: &Node) -> bool {
        let mut current = node.parent;
        if std::ptr::eq(node, &self.nodes[self.root]) {
            return true;
        }
        while let Some(id) = current {
            if id == self.root {
                return true;
            }
            current = self.nodes[id].parent;
        }
        false
    }
}
